.pipelines/templates/.builder-release-template-windows.yaml

```diff
@@ -4,11 +4,37 @@ parameters:
       default: 2019
 
 steps:
+# For Gen 2, if we want to use sig for test purpose only, we set the pipeline variables
+#               SIG_GALLERY_NAME, SIG_IMAGE_NAME_PREFIX, and SIG_IMAGE_VERSION.
+#               SIG_IMAGE_NAME_PREFIX will be used to define the task variable SIG_IMAGE_NAME for
+#               the packer destination SIG.
+#               SIG_IMAGE_NAME, SIG_GALLERY_NAME, and SIG_IMAGE_VERSION will be passed below.
+#               Task variable GEN2_SIG_FOR_PRODUCTION is set to False and passed to the following steps.
+#               Built sig will not be deleted because it is for future testing.
+#            if we want to publish the VHD, instead of using pipeline variables,
+#               we will hard-code the task variables SIG_GALLERY_NAME, SIG_IMAGE_NAME, and SIG_IMAGE_VERSION.
+#               Task variable GEN2_SIG_FOR_PRODUCTION is set to True and passed to the following steps.
+#               Built sig will be deleted because it has been converted to VHD, and thus not needed.
   - bash: |
-      if [[ -n ${SIG_GALLERY_NAME} && -n ${SIG_IMAGE_NAME_PREFIX} && -n ${SIG_IMAGE_VERSION} ]]; then m="sigMode"; sigImageName="${SIG_IMAGE_NAME_PREFIX}-${WINDOWS_SKU}"; else m="default"; fi && \
-      echo "Set SIG_IMAGE_NAME to $sigImageName" && \
-      echo "##vso[task.setvariable variable=SIG_IMAGE_NAME]$sigImageName" && \
+      if [[ $(HYPERV_GENERATION) == "V2" ]]; then \
+        m="sigMode"; \
+        if [[ -n ${SIG_GALLERY_NAME} && -n ${SIG_IMAGE_NAME_PREFIX} && -n ${SIG_IMAGE_VERSION} ]]; then \   
+          sigImageName="${SIG_IMAGE_NAME_PREFIX}-${WINDOWS_SKU}-gen2"; \
+          echo "##vso[task.setvariable variable=GEN2_SIG_FOR_PRODUCTION]False"; \
+        else
+          sigImageName="windows-${WINDOWS_SKU}"; \
+          echo "##vso[task.setvariable variable=SIG_GALLERY_NAME]WS2022Gen2Gallery_${RANDOM}"; \
```

Comment:
Use `WS2022Gen2Gallery-${RANDOM}` instead of `WS2022Gen2Gallery_${RANDOM}`

Reply:
I tried to use dash before. The log says **"ERROR: (BadRequest) Gallery name is invalid. Allowed characters are English alphanumeric characters, with underscores and periods allowed in the middle, up to 80 characters total. All other special characters, including dashes, are disallowed.".** 

See the most recent pipeline [https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=59422027&view=logs&j=05b0285b-7424-5da1-4fe4-29b48f7b1d3a&t=4f5de716-8c9d-57d8-0c59-15d673379c58](url)

Reply:
I see. Maybe we can use "WS2022Gen2Gallery{RANDOM}". I am not sure whether it is better to use "WS2022Gen2Gallery$(Build.BuildNumber)".

Reply:
Maybe $(Build.BuildNumber) is more informative since the BuildNumber like `20220823.5_merge_****` tells the date? When we check the gallery on portal to see if the gallery has been deleted - sometimes undeleted resources will break the pipeline, one can trace back to the code commit on that day and see what went wrong???

Further, I think we can use something similar to `STORAGE_ACCOUNT_NAME="aksimages${CREATE_TIME}$RANDOM"`, i.e. `WS2022Gen2Gallery$(Build.BuildNumber)${RANDOM}`?

Reply:
storageaccount has length limitation "must be between 3 and 24 characters in length and may contain numbers and lowercase letters only."

Reply:
I meant we can emulate `STORAGE_ACCOUNT_NAME` to set `SIG_GALLERY_NAME` <= `WS2022Gen2Gallery$(Build.BuildNumber)${RANDOM}`. Gallery name is "up to 80 characters total".

**Proposal**: `WS2022Gen2Gallery$(Build.BuildNumber)${RANDOM}`?

Reply:
I see. WS2022Gen2Gallery$(Build.BuildNumber)${RANDOM} LGTM